### ISSUE TITLE
feat(compute): native tree-attention kernels for Vulkan and Metal with arbitrary tree-mask support

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12724]
# feat(compute): native tree-attention kernels for Vulkan and Metal with arbitrary tree-mask support

```routing
lane: internal/compute
paths: internal/compute/spec_verify.go, internal/compute/tree_attention.go, internal/compute/vulkan_ops_vulkan.go
expected_steps: 4
```

## Lane

`internal/compute` — hardware compute HAL, attention kernels, and tensor math execution.

## Problem frame

- Centrality: Core
- Portfolio tier: 2 (serving)
- Priority: P1
- P1: advanced - eliminates CPU reference fallback during speculative candidate tree verification on Apple Silicon and AMD Vulkan
- P2: preserved - maintains bit-level equivalence and causal masking invariants for standard linear speculative sequences
- P3: advanced - enables arbitrary DAG and tree attention masks required by EAGLE-3 and multi-branch speculative decoders
- P4: preserved - zero allocation overhead during attention execution with register-resident accumulators

## Parent context

Part of the Native Speculative Acceleration Series (`docs/tickets/native-speculative-acceleration/INDEX.md`), establishing hardware-accelerated speculative verification primitives across Vulkan and Metal backends.

## Current state

In `internal/compute/spec_verify.go`, `SpecVerifyAttention` implements a split-KV speculative verify attention kernel, but its reference logic (`specVerifyAttentionHost`) strictly enforces a linear causal sequence (`attendLen := prefix + qi + 1`). When evaluating speculative candidate trees (where draft tokens branch into alternative paths sharing ancestor prefixes), this linear constraint fails because sibling draft branches must not attend to each other.

Furthermore, on both Vulkan (`internal/compute/vulkan_ops_vulkan.go:108`) and Metal (`internal/compute/metal.go:690`), `SpecVerifyAttention` currently falls back directly to the CPU reference implementation (`ref.SpecVerifyAttention(...)`), forcing expensive GPU-to-CPU synchronization, host DRAM readbacks, and high dispatch latency during verification steps.

## Why this is next

Speculative tree decoding (such as EAGLE-3 or tree n-gram search) evaluates 16 to 32 candidate tokens simultaneously in a single target model forward pass. Falling back to the CPU reference on Vulkan (AMD Strix Halo / Linux) and Metal (Apple Silicon) negates speculative throughput gains due to device synchronizations. Providing native GPU tree-attention kernels with arbitrary mask support unlocks true 2x to 3x decode acceleration on local accelerator hardware.

## Working spine

`SpecVerifyAttentionBackend` extended with `TreeSpecVerifyAttention` -> CPU reference validates arbitrary 2D boolean tree-attention masks with bit-identical outputs -> native Metal compute kernel executes tree attention using threadgroup shared memory and SIMDgroup operations -> native Vulkan compute shader evaluates tree attention via workgroup memory and subgroup instructions -> verified multi-token tree verification executes entirely on device without CPU round-trips.

## Core through-line

1. Define `TreeSpecVerifyAttentionBackend` and `TreeSpecVerifyAttention(q, k, v, mask, out *Tensor, qLen, kvLen, nH, nHkv, d int) error` in `internal/compute/tree_attention.go`, accepting a 2D tree attention adjacency mask of shape `[qLen, qLen]` (or `[qLen, kvLen]`).
2. Author byte-accurate CPU reference implementation in `internal/compute/tree_attention.go` that masks out non-ancestor candidate tokens while preserving complete visibility over the historical KV prefix (`kvLen - qLen`).
3. Author native Metal MSL shader in `internal/compute/metal.go` utilizing threadgroup shared memory to tile queries and keys, applying the tree mask in-flight during the online-softmax score reduction.
4. Author native Vulkan GLSL/SPIR-V shader in `internal/compute/vulkan_ops_vulkan.go` implementing the masked attention reduction using Vulkan subgroup operations without dynamic heap allocation.
5. Provide unit and regression tests asserting cosine similarity >= 0.9999 against CPU reference and verifying strict causal isolation between sibling candidate branches.

## Gold-plating boundary

Strictly excluded from this work unit:
- Dynamic beam-search tree pruning heuristics outside the compute kernel interface.
- Sparse FlashAttention block-sparse layouts (tree attention applies to small $qLen \le 64$ dense draft matrices).
- CUDA FlashInfer tree-attention integration (owned by existing CUDA kernel paths).

## Verifiable Witness

```bash
go test -v ./internal/compute -run 'TestTreeSpecVerifyAttention' -count=1
```

## Witness

```bash
go test -v ./internal/compute -run 'TestTreeSpecVerifyAttention' -count=1
```

### File:Line Seam

- `internal/compute/spec_verify.go:10 (SpecVerifyBackend)`
- `internal/compute/vulkan_ops_vulkan.go:108 (vulkanBackend.SpecVerifyAttention)`
- `internal/compute/metal.go:690 (metalBackend.SpecVerifyAttention)`
- `internal/compute/tree_attention.go:1 (TreeSpecVerifyAttention)`

### Blast Radius & Affected Lanes

- **Primary Lane:** `internal/compute`.
- **Unaffected Lanes:** `internal/model`, `internal/agent`, `platform/*` remain unaffected.

### Quarantined Fallback Mechanism

When optional tree-attention device shader compilation is unavailable or fails validation, `TreeSpecVerifyAttention` gracefully falls back to the CPU reference with an explicit logged debug trace; all existing linear `SpecVerifyAttention` call sites remain backward compatible `[SW-VERIFIED]`.

## Scoped Acceptance Criteria

- [ ] 1. [SW-VERIFIED] Implement `TreeSpecVerifyAttention` CPU reference in `internal/compute/tree_attention.go` supporting arbitrary 2D candidate tree masks with cosine similarity >= 0.9999 against unmasked causal baseline.
- [ ] 2. [SW-VERIFIED] Unit tests verify sibling candidate branches in a tree cannot attend to each other and only attend to true ancestors and prefix KV tokens.
- [ ] 3. [SW-VERIFIED] Implement native Metal tree-attention compute pipeline in `internal/compute/metal.go` eliminating CPU fallback on macOS Metal.
- [ ] 4. [SW-VERIFIED] Implement native Vulkan tree-attention compute pipeline in `internal/compute/vulkan_ops_vulkan.go` eliminating CPU fallback on Vulkan.
- [ ] 5. [HW-WITNESSED] Verify on physical GPU hardware (Vulkan or Metal) that a 16-token candidate tree verification executes with zero host memory round-trips and sub-1ms kernel execution latency.

## Done condition

- [ ] 1. [SW-VERIFIED] Implement `TreeSpecVerifyAttention` CPU reference in `internal/compute/tree_attention.go` supporting arbitrary 2D candidate tree masks with cosine similarity >= 0.9999 against unmasked causal baseline.
- [ ] 2. [SW-VERIFIED] Unit tests verify sibling candidate branches in a tree cannot attend to each other and only attend to true ancestors and prefix KV tokens.
- [ ] 3. [SW-VERIFIED] Implement native Metal tree-attention compute pipeline in `internal/compute/metal.go` eliminating CPU fallback on macOS Metal.
- [ ] 4. [SW-VERIFIED] Implement native Vulkan tree-attention compute pipeline in `internal/compute/vulkan_ops_vulkan.go` eliminating CPU fallback on Vulkan.
- [ ] 5. [HW-WITNESSED] Verify on physical GPU hardware (Vulkan or Metal) that a 16-token candidate tree verification executes with zero host memory round-trips and sub-1ms kernel execution latency.

## Acceptance gate

`go test -v ./internal/compute -run 'TestTreeSpecVerifyAttention' -count=1` exits 0 with 0 boundary violations.

## Closure binding

Resolving commit carries `(fak compute)` trailer.

## Likely files

- `internal/compute/tree_attention.go`
- `internal/compute/tree_attention_test.go`
- `internal/compute/spec_verify.go`
- `internal/compute/vulkan_ops_vulkan.go`
- `internal/compute/metal.go`

## Expected steps

4
